sandbox/cgroup/monitor_test.go

```diff
@@ -43,125 +46,211 @@ type monitorSuite struct {
 var _ = Suite(&monitorSuite{})
 
 func (s *monitorSuite) SetUpTest(c *C) {
+	logger.SimpleSetup()
 	s.BaseTest.SetUpTest(c)
 
 	dirs.SetRootDir(c.MkDir())
 	s.AddCleanup(func() { dirs.SetRootDir("/") })
 
+	s.tempDir = c.MkDir()
+
 	s.eventsCh = make(chan string)
 	s.AddCleanup(func() { close(s.eventsCh) })
-
-	s.calibrateInotifyDelay(c)
 }
 
-func (s *monitorSuite) calibrateInotifyDelay(c *C) {
-	folder1 := s.makeTestFolder(c, "folder1")
-	filelist := []string{folder1}
-	err := cgroup.MonitorDelete(filelist, "test1", s.eventsCh)
-	c.Assert(err, IsNil)
-	var start time.Time
-	go func() {
-		start = time.Now()
-		err = os.Remove(folder1)
-		c.Assert(err, IsNil)
-	}()
-	<-s.eventsCh
-	d := time.Now().Sub(start)
-	// On a modern machine the duration "d" for inotify delivery is
-	// around 30-100µs so even the very conservative multiplication means
-	// the delay is typically 3ms-10ms.
-	s.inotifyWait = 100 * d
-	switch {
-	case s.inotifyWait > 1*time.Second:
-		s.inotifyWait = 1 * time.Second
-	case s.inotifyWait < 10*time.Millisecond:
-		s.inotifyWait = 10 * time.Millisecond
-	}
-}
-
-func (s *monitorSuite) makeTestFolder(c *C, name string) (fullPath string) {
-	fullPath = path.Join(c.MkDir(), name)
+func makeTestFolder(c *C, root string, name string) (fullPath string) {
+	fullPath = path.Join(root, name)
 	err := os.Mkdir(fullPath, 0755)
 	c.Assert(err, IsNil)
 	return fullPath
 }
 
 func (s *monitorSuite) TestMonitorSnapBasicWork(c *C) {
-	folder1 := s.makeTestFolder(c, "folder1")
-	folder2 := s.makeTestFolder(c, "folder2")
+	cb := 0
+	monitorAdded := make(chan struct{})
+	ctx, cancel := context.WithCancel(context.Background())
+	s.AddCleanup(cgroup.MockInotifyWatcher(ctx, func(w *cgroup.InotifyWatcher, name string) {
+		cb++
+		switch cb {
+		case 1:
+			if cb == 1 {
+				close(monitorAdded)
+			}
+			c.Check(name, Equals, "test1")
+		case 2, 3:
+		default:
+			c.Fatalf("unexpected callback: %v %q", cb, name)
+		}
+
+	}))
+	defer cgroup.MockInitWatcherClose()
+	defer cancel()
```

Comment:
Since the watcher has a cancel context inside anyway, is this another layer necessary? Isn't calling Close just canceling the context inside?

Reply:
good point, I forgot to drop this after adding Close()